cmd/kubeadm/app/cmd/join.go

```diff
@@ -235,34 +279,228 @@ func (j *Join) Validate(cmd *cobra.Command) error {
+	}
+
+	// Converts public API to internal API
+	masterConfiguration := &kubeadmapi.MasterConfiguration{}
+	legacyscheme.Scheme.Convert(masterConfigurationExt, masterConfiguration, nil)
 
-	// Write the bootstrap kubelet config file or the TLS-Boostrapped kubelet config file down to disk
-	if err := kubeconfigutil.WriteToDisk(kubeconfigFile, cfg); err != nil {
+	return masterConfiguration, nil
+}
+
+// CheckIfReadyForJoinAsMaster ensures that the cluster is in a configuration that supports
+// joining additional masters
+func (j *Join) CheckIfReadyForJoinAsMaster(masterConfiguration *kubeadmapi.MasterConfiguration) error {
+
+	// checks if the cluster was created with features.HighAvailability enabled
+	if !features.Enabled(masterConfiguration.FeatureGates, features.HighAvailability) {
+		return fmt.Errorf("unable to join a new master node on a cluster that wasn't created without '--feature-gates=HighAvailability=true'")
+	}
+
+	// ** Note for reviewers:
+	// ** If the cluster was initialized with --feature-gates=HighAvailability=true following checks were already executed
+	// ** during kubeadm init:
+	// ** 1. the cluster uses an external etcd
+	// ** 2. the cluster uses an external load balancer
+	// ** Should we double check the above assumptions before joining?
+
+	// if the certificates for the control plane are not stored in secrets (if the certificates are stored in a local pki folder),
+	if !features.Enabled(masterConfiguration.FeatureGates, features.StoreCertsInSecrets) {
+
+		// checks if the certificates for the control plane were already provided by the user in
+		// the local pki folder and checks if those certs are compliant with the requirements
+		// ** Note for the reviewers:
+		// ** see notes in phases/cert about how sharing validate cert functions between the two features
+		err := certsphase.CheckIfLocalPkiReadyForJoinAsMaster(masterConfiguration)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// PrepareForJoinAsMaster makes all preparation activities require for a node joining as a master
+func (j *Join) PrepareForJoinAsMaster(masterConfiguration *kubeadmapi.MasterConfiguration) error {
+
+	// if the certificates for the control plane are stored in a local pki folder
+	// (if certificates are not stored in secrets)
+	if !features.Enabled(masterConfiguration.FeatureGates, features.StoreCertsInSecrets) {
+
+		// Creates the missing certificates
+		// NB. This function will use the certificate authority and service account signing key
+		//     provided by the user before executing init --master.
+		//     (those certificates/keys must be the same across all master instances)
+		certsphase.CreatePKIAssets(masterConfiguration)
+	}
+
+	// if the cluster uses StaticPods manifests (if the cluster is not self hosted)
+	if !features.Enabled(masterConfiguration.FeatureGates, features.SelfHosting) {
+
+		// Generate kubeconfig files for controller manager, scheduler and for the admin/kubeadm itself
```

Comment:
Do we even want to allow non-self hosted HA, when we already have instructions we can point to?  It seems like a mgmt issue.

Reply:
Let's discuss this in next call. Two main consideration from my side:
- With `kubeadm join --master` supporting non-self hosted HA we can further simplify the kubeadm HA tutorial and as a consequence we will reduce the possible problems the user can face
- IMO we should provide support for non-self hosted HA until we address the checkpointing issue and self-hosting became replaces static pods as a default deployment mode for kubeadm.

(reminder) if the final decision will be to support only HA with self-hosting, we should check that self-hosting is on when doing `kubeadm init --feature-gates=HighAvailability`